### PR TITLE
DBZ-6076: Add support for custom SourceInfoStructMaker for adding new fields to source field

### PR DESCRIPTION
**Pull request by adasari**
Added support to use custom connector config using custom task.

### CODE REVIEW

**Comment by jpechane** [comment]
@adasari Hi, I converted the PR to the draft as I suppose this is just a first step, not the actual implementation. Am I right?

**Comment by jpechane** [comment]
@adasari OK, in this case it would be better to add a new config option that would containe the class name of the custom structmaker and if present then instantiate it and pass the connector config to it.
It would be similar to what we do with `topic.naming.strategy`.

**Comment by adasari** [comment]
@jpechane Please review the changes. i will raise PRs to other repos when these changes are ok.

**Comment by adasari** [comment]
Ping

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 1]
```diff
@@ -693,7 +700,8 @@ public static FieldNameAdjustmentMode parse(String value) {
                     SIGNAL_ENABLED_CHANNELS,
                     TOPIC_NAMING_STRATEGY,
                     NOTIFICATION_ENABLED_CHANNELS,
-                    SinkNotificationChannel.NOTIFICATION_TOPIC)
+                    SinkNotificationChannel.NOTIFICATION_TOPIC,
+                    SOURCE_INFO_STRUCT_MAKER)
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:704]: @adasari This should be included in per-connector config class as the definition changes. Othrwise the tools will not see the default value in config definition.

**Review by jpechane — COMMENTED** [review]
@adasari LGTM, I kept only one comment.
Please add a test to at least one connector to verify that this works as expected.
At the same time I beleive you can prepr PR's to non-core repo connectors as I don't expect any further changes.

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java (hunk 1 above)
Comment (adasari) [reply debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:704]: Left a comment - https://github.com/debezium/debezium/pull/4258/files#r1213185645. thanks.

**Comment by adasari** [comment]
Added test to postgres connector.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 2]
```diff
@@ -456,7 +460,7 @@ protected boolean supportsSchemaChangesDuringIncrementalSnapshot() {
 
     @Override
     protected SourceInfoStructMaker<? extends AbstractSourceInfo> getSourceInfoStructMaker(Version version) {
-        return new SqlServerSourceInfoStructMaker(Module.name(), Module.version(), this);
+        return getSourceInfoStructMaker(SqlServerConnectorConfig.SOURCE_INFO_STRUCT_MAKER, Module.name(), Module.version(), this);
```
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java:463]: @adasari I understand your point. The issue is that every connector has/should have a different default value and tolling like UI should be able to show that default value. And as the Field is static I don't see any way how to configure it per-connector to have a distinct default value and consume it in CommonConnectorConfig.

**Comment by jpechane** [comment]
@adasari Applied, thanks a lot!

**Comment by zackary-naas** [comment]
> [@jpechane](/jpechane) [jpechane](/jpechane) [force-pushed](/debezium/debezium/compare/28d04306a0805e69a501d61ffe9b3466e6aad5c4..f5b91dd701e40954e5ee875cb3c40cd4b2048a82) the DBZ-6076 branch 2 times, most recently from [ `28d0430` ](/debezium/debezium/commit/28d04306a0805e69a501d61ffe9b3466e6aad5c4) to [ `f5b91dd` ](/debezium/debezium/commit/f5b91dd701e40954e5ee875cb3c40cd4b2048a82) [ Compare ](/debezium/debezium/compare/28d04306a0805e69a501d61ffe9b3466e6aad5c4..f5b91dd701e40954e5ee875cb3c40cd4b2048a82) [ June 14, 2023 23:11 ](#event-9536053121)
> [ ](/adasari)

@adasari how did you implement writing to this custom SourceInfo field as the connector handles events?

**Comment by adasari** [comment]
> > [@jpechane](/jpechane) [jpechane](/jpechane) [force-pushed](/debezium/debezium/compare/28d04306a0805e69a501d61ffe9b3466e6aad5c4..f5b91dd701e40954e5ee875cb3c40cd4b2048a82) the DBZ-6076 branch 2 times, most recently from [ `28d0430` ](/debezium/debezium/commit/28d04306a0805e69a501d61ffe9b3466e6aad5c4) to [ `f5b91dd` ](/debezium/debezium/commit/f5b91dd701e40954e5ee875cb3c40cd4b2048a82) [ Compare ](/debezium/debezium/compare/28d04306a0805e69a501d61ffe9b3466e6aad5c4..f5b91dd701e40954e5ee875cb3c40cd4b2048a82) [ June 14, 2023 23:11 ](#event-9536053121)
> > [ ](/adasari)
> 
> @adasari how did you implement writing to this custom SourceInfo field as the connector handles events?

By custom SourceInfoStructMaker. Reference implementation is https://github.com/debezium/debezium/blob/7d6b4af1dea3e7b0cb6e9abbc590427952814de6/debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/CustomPostgresSourceInfoStructMaker.java#L19

**Comment by zackary-naas** [comment]
Apologies, maybe I misunderstood what you were using this for. To explain a bit more what I would like to do - there's an additional field that shows up in the MySQL binlog data, and I'd like to extract that into the SourceInfo at the time of handling an event. As far as I can tell, there is not a way to do that.

To confirm and say it another way - what you are showing with the reference implementation is how to add a static field to the source info correct?

**Comment by adasari** [comment]
You can use it to add static field or dynamic field based on other SourceInfo fields. PostgresSourceInfoStructMaker#struct has SourceInfo only.

**Comment by adasari** [comment]
> Apologies, maybe I misunderstood what you were using this for. To explain a bit more what I would like to do - there's an additional field that shows up in the MySQL binlog data, and I'd like to extract that into the SourceInfo at the time of handling an event. As far as I can tell, there is not a way to do that.
> 
> To confirm and say it another way - what you are showing with the reference implementation is how to add a static field to the source info correct?

Q:: is bin log data is present in SourceRecord ?  
Implementation ideas:
1. Use custom SourceInfoStructMaker if new field of source info is either static or dynamic based on other source info fields.
2. Use SMT if it is  new field is already part of SourceRecord
3.  SMT can be overkill if you regenerate the schema. If SMT is a overkill in your case, Go with hybrid approach.
    a. Use custom SourceInfoStructMaker to add the field to source info schema and default value (like -1 etc)
    b. Use SMT or custom ConnectorTask of doPoll () implementation that fills new field from SourceRecord fields.

Please evaluate all the options and select that is suitable for your use case.

**Pull request merged**